source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,684 @@
+/++
+This module helps building functions from other functions.
+
+ It is sometimes necessary to create a function which is an exact copy of
+ another function. Or sometimes it is necessary to introduce a few variations,
+ while carrying all the other aspects. Because of function attributes,
+ parameter storage classes and user-defined attributes, this requires building
+ a string mixin. In addition, the mixed-in code must refer only to local names,
+ if it is to work across module boundaires. This problem and its solution are
+ by Adam D. Ruppe in a Tip of the Week, available here:
+ https://stackoverflow.com/questions/32615733/struct-composition-with-mixin-and-templates/32621854#32621854
+
+ This module facilitates the  creation of such mixins.
+
+ +/
+
+module bolts.experimental.refraction;
+
+import std.array;
+import std.format;
+import std.meta;
+import std.algorithm.iteration : map;
+import std.range : iota;
+import std.traits : functionAttributes, FunctionAttribute;
+
+// Do not require caller module to import 'std.traits'. Instead use our own
+// aliases in mixtures.
+alias ReturnType = std.traits.ReturnType;
+alias Parameters = std.traits.Parameters;
+
+template ParameterAttribute(alias F, int i, int j) {
+    static if (is(typeof(F) P == __parameters)) {
+        alias ParameterAttribute = Alias!(__traits(getAttributes, P[i..i+1])[j]);
+    }
+}
+
+unittest {
+    struct virtual;
+    void kick(int times, @virtual @("Animal") Object animal);
+    static assert(is(ParameterAttribute!(kick, 1, 0) == virtual));
+    static assert(ParameterAttribute!(kick, 1, 1) == "Animal");
+}
+
+/**
+   Return a `Function` object that captures all the aspects of `fun`, using the
+   value of `localName` to represent the return and parameter types, and the
+   UDAs. Set the `Function`'s `index` property to `index`, or to -1 if it is
+   not specified.
+
+   The `localName` parameter is, in general, *not* the function name. Rather,
+   it is a compile-time expression that involves only symbols that exist in the
+   caller's scope, for example a function alias passed as a template
+   parameter. See
+   https://stackoverflow.com/questions/32615733/struct-composition-with-mixin-and-templates/32621854#32621854
+   for a detailed explanation.
+
+   Params:
+   fun = a function
+   localName = a string that represents `fun` in the caller's context
```

Comment:
index docs missing

Reply:
Funny github does not allow me to reply to your message below. Putting my comments here.

> Doc fix. No withIndex anymore.

...and the doc of `index` needed updating as well, for the 2nd `refract` is now `functionsOf`.

> Also could still rename to overloadIndex?

So I'm pretty sure that half of the time it will be more useful to have the index of the function in the module or aggregate. For example, suppose we wanted to implement the `Mock` example fully. It would be much more convenient to have the flat index. Assume for a moment that `functionsOf` stores the flat index, not the overload's, in `index`. We could write:

```dlang
unittest {
    interface GrandTour {
        pure int foo();
        @nogc @trusted nothrow ref int foo(out real, return ref int, lazy int);
        @safe scope void bar(scope Object);
    }

    class Mock(Interface) : Interface {
        enum functions = functionsOf!(Interface, "Interface");
        int[functions.length] calls;
        static foreach (Function fun; functions) {
            mixin({
                    if (is(mixin(fun.returnType) == void)) {
                        return fun.withBody("{ ++calls[fun.index]; }").mixture;
                    } else if (fun.attributes & FunctionAttribute.ref_) {
                        return fun.withBody(q{{
                                    ++calls[fun.index];
                                    static %s rv;
                                    return rv;
                                }}.format(fun.returnType)).mixture;
                    } else {
                        return fun.withBody(q{{
                                    ++calls[fun.index];
                                    return %s.init;
                                }}.format(fun.returnType)).mixture;
                    }
                }());
        }
    }

    auto mock = new Mock!GrandTour;
    real x;
    int i, l;
    mock.foo(x, i, l++) = 1;
    assert(mock.foo(x, i, l++) == 1);
    assert(mock.calls[1] == 1);
    assert(l == 0);
}
```

Of course this is a bit naive - I had to make all functions mutable. But you get the idea...

So, this morning I realized that we should store both: `index` which contains the flat index, and - you'll be happy - `overloadIndex` which is what we have now, renamed.

I am going to put this change in a separate commit, in case we can't agree on it.

Reply:
OK forget about this "other" index. While playing with this, I realized that:
* the position a function (overload) inside a scope is...a complicated concept. `__traits` returns functions by name, then by overload.
* all I need is some sort of flat index, and the position of the `Function` inside the array returned by `functionsOf` does just fine.

So I am going to just rename `index` to `overloadIndex` as you suggest :)

```dlang
// dmd -i -I=~/dev/bolts/source/ -main -unittest -mixin=mixins.d refraction_demo.d

module demo;

import std.format;
import std.traits : FunctionAttribute;
import bolts.experimental.refraction;
import std.algorithm.searching;

unittest {
    interface GrandTour {
        pure int foo();
        @nogc @trusted nothrow ref int foo(out real, return ref int, lazy int);
        @safe scope void bar(scope Object);
    }

    class Mock(Interface) : Interface {
        enum functions = functionsOf!(Interface, "Interface");
        int[functions.length] calls;
        static foreach (Function fun; functions) {
            mixin({
                    auto index = functions.countUntil(fun);
                    if (is(mixin(fun.returnType) == void)) {
                        return fun.withBody("{ ++calls[%s]; }").mixture.format(index);
                    } else if (fun.attributes & FunctionAttribute.ref_) {
                        return fun.withBody(q{{
                                    ++calls[%s];
                                    static %s rv;
                                    return rv;
                                }}.format(index, fun.returnType)).mixture;
                    } else {
                        return fun.withBody(q{{
                                    ++calls[%s];
                                    return %s.init;
                                }}.format(index, fun.returnType)).mixture;
                    }
                }());
        }
    }

    auto mock = new Mock!GrandTour;
    real x;
    int i, l;
    mock.foo(x, i, l++) = 1;
    assert(mock.foo(x, i, l++) == 1);
    assert(mock.calls[1] == 1);
    assert(l == 0);
}
```